File: frontend/src/lib/types/providers.ts
import type { ModelPricing } from "@/lib/types/chat";
import type { UUID } from "@/lib/types/common";

/** Mirrors `app/schemas/enums.py::ProviderType`. */
export type ProviderType = "openrouter" | "ollama" | "pinecone";

/** Mirrors `app/schemas/enums.py::ProviderKind`. */
export type ProviderKind = "embedding" | "chat" | "vector_store";

/** Mirrors `app/schemas/providers.py::ConfigFieldKind`. */
export type ProviderConfigFieldKind = "string" | "secret" | "url";

/** Mirrors `app/schemas/providers.py::ProviderConfigField`. */
export interface ProviderConfigField {
  name: string;
  label: string;
  kind: ProviderConfigFieldKind;
  required: boolean;
  placeholder?: string | null;
  description?: string | null;
}

/** Mirrors `app/schemas/providers.py::ProviderTypeRead` (`GET /api/providers`). */
export interface ProviderTypeInfo {
  provider_type: string;
  label: string;
  kinds: ProviderKind[];
  config_fields: ProviderConfigField[];
  docs_url?: string | null;
  max_connections_per_user?: number | null;
  recommended: boolean;
  builtin: boolean;
  available: boolean;
}

/** Mirrors `app/schemas/providers.py::ConnectionRead` — secrets never included. */
export interface ProviderConnection {
  id: UUID;
  provider_type: ProviderType;
  label: string;
  kinds: ProviderKind[];
  config: Record<string, string>;
  secrets_configured: Record<string, boolean>;
  created_at: string;
  updated_at: string;
}

/** Mirrors `app/schemas/providers.py::ConnectionCreate`. */
export interface ConnectionCreateRequest {
  provider_type: string;
  label: string;
  config: Record<string, string>;
}

/** Mirrors `app/schemas/providers.py::ConnectionUpdate`. */
export interface ConnectionUpdateRequest {
  label?: string;
  config?: Record<string, string>;
}

/** Mirrors `app/schemas/providers.py::ConnectionValidationResult`. */
export interface ConnectionValidationResult {
  valid: boolean;
  message?: string | null;
}

/** Mirrors `app/schemas/providers.py::CatalogModel` — a model qualified by its connection. */
export interface CatalogModel {
  connection_id: UUID;
  connection_label: string;
  provider_type: ProviderType;
  id: string;
  name: string;
  description?: string | null;
  context_length?: number | null;
  pricing?: ModelPricing | null;
  dimension?: number | null;
  supported_parameters: string[];
  default_parameters?: Record<string, unknown> | null;
}

/** Mirrors `app/schemas/providers.py::ConnectionCatalogError`. */
export interface ConnectionCatalogError {
  connection_id: UUID;
  connection_label: string;
  message: string;
}

/** Mirrors `app/schemas/providers.py::CatalogMetadata`. */
export interface CatalogMetadata {
  freshness: "fresh" | "stale";
  age_seconds: number;
  refreshing: boolean;
  warning?: string | null;
}

/** Mirrors `app/schemas/providers.py::ModelCatalogResponse` (`GET /api/models`). */
export interface ModelCatalogResponse {
  models: CatalogModel[];
  connection_errors: ConnectionCatalogError[];
  meta: CatalogMetadata;
}

/** Mirrors `app/schemas/providers.py::EmbeddingDimensionResponse`. */
export interface EmbeddingDimensionResponse {
  connection_id: UUID;
  model_id: string;
  dimension?: number | null;
}

File: app/schemas/providers.py
"""Wire contract for provider connections and the provider-type catalog.

The per-type connection config models (`OpenRouterConnectionConfig`, ...) are
the single validation point for what lands in `provider_connections.config`;
the connections service validates through them before writing, and adapters
read through them at construction time. `ConnectionRead` never carries secret
values — secret fields are echoed as `secrets_configured` booleans only.
"""

from __future__ import annotations

from datetime import datetime
from enum import StrEnum
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.schemas.base import DateTimeConfigMixin
from app.schemas.enums import ProviderKind, ProviderType
from app.schemas.models import ModelPricing


class ConfigFieldKind(StrEnum):
    """Rendering kinds for provider config fields (drives the generic form)."""

    STRING = "string"
    SECRET = "secret"
    URL = "url"


class ProviderConfigField(BaseModel):
    """One field of a provider type's connection config, for form rendering."""

    name: str
    label: str
    kind: ConfigFieldKind
    required: bool = True
    placeholder: str | None = None
    description: str | None = None


class ProviderTypeRead(BaseModel):
    """One entry of the provider-type catalog (`GET /api/providers`).

    `builtin` entries (pgvector) need no connection; `available` reports
    whether a builtin is usable on this deployment.
    """

    provider_type: str
    label: str
    kinds: list[ProviderKind]
    config_fields: list[ProviderConfigField]
    docs_url: str | None = None
    max_connections_per_user: int | None = None
    recommended: bool = False
    builtin: bool = False
    available: bool = True


class OpenRouterConnectionConfig(BaseModel):
    """Stored config for an OpenRouter connection."""

    api_key: str = Field(min_length=1)


class OllamaConnectionConfig(BaseModel):
    """Stored config for an Ollama connection."""

    base_url: str = Field(min_length=1)
    api_key: str | None = None

    @field_validator("base_url")
    @classmethod
    def normalize_base_url(cls, value: str) -> str:
        """Require an http(s) URL and strip the trailing slash."""
        cleaned = value.strip().rstrip("/")
        if not cleaned.startswith(("http://", "https://")):
            raise ValueError("Base URL must start with http:// or https://.")
        return cleaned


class PineconeConnectionConfig(BaseModel):
    """Stored config for a Pinecone connection."""

    api_key: str = Field(min_length=1)


class ConnectionCreate(BaseModel):
    """Payload for registering a provider connection."""

    provider_type: ProviderType
    label: str = Field(min_length=1, max_length=100)
    config: dict[str, Any]


class ConnectionUpdate(BaseModel):
    """Payload for editing a connection.

    `config` is a partial overlay: only the provided fields replace stored
    values, so relabeling never requires re-entering secrets.
    """

    label: str | None = Field(default=None, min_length=1, max_length=100)
    config: dict[str, Any] | None = None


class ConnectionRead(DateTimeConfigMixin, BaseModel):
    """A connection as returned to clients — secret values never included."""

    model_config = ConfigDict(**DateTimeConfigMixin.model_config)

    id: UUID
    provider_type: ProviderType
    label: str
    kinds: list[ProviderKind]
    config: dict[str, str]
    secrets_configured: dict[str, bool]
    created_at: datetime
    updated_at: datetime


class ConnectionValidateRequest(BaseModel):
    """An unsaved connection config to probe before creating it."""

    provider_type: ProviderType
    config: dict[str, Any]


class ConnectionValidationResult(BaseModel):
    """Outcome of probing a connection's credentials/reachability."""

    valid: bool
    message: str | None = None


class CatalogModel(BaseModel):
    """One selectable model qualified by the connection that serves it."""

    connection_id: UUID
    connection_label: str
    provider_type: ProviderType
    id: str
    name: str
    description: str | None = None
    context_length: int | None = None
    max_input_tokens: int | None = None
    pricing: ModelPricing | None = None
    dimension: int | None = None
    supported_parameters: list[str] = Field(default_factory=list)
    default_parameters: dict[str, Any] | None = None


class ConnectionCatalogError(BaseModel):
    """A connection whose catalog fetch failed while listing models."""

    connection_id: UUID
    connection_label: str
    message: str


class CatalogMetadata(BaseModel):
    """Freshness of the unified catalog returned to model selectors."""

    freshness: Literal["fresh", "stale"] = "fresh"
    age_seconds: float = Field(default=0, ge=0)
    refreshing: bool = False
    warning: str | None = None


class ModelCatalogResponse(BaseModel):
    """Unified model listing across every connection of the requested kind.

    One unreachable connection degrades to a `connection_errors` entry rather
    than failing the whole listing.
    """

    models: list[CatalogModel] = Field(default_factory=list)
    connection_errors: list[ConnectionCatalogError] = Field(default_factory=list)
    meta: CatalogMetadata = Field(default_factory=CatalogMetadata)


class EmbeddingDimensionResponse(BaseModel):
    """Dimension lookup qualified by the exact provider connection and model."""

    connection_id: UUID
    model_id: str
    dimension: int | None


class ProviderCoverage(BaseModel):
    """Which provider kinds the user's connections (plus builtins) cover."""

    has_embedding: bool
    has_chat: bool
    has_vector_store: bool
